src/main/java/network/brightspots/rcv/HartCvrReader.java

```diff
@@ -0,0 +1,169 @@
+    for (File child : children) {
+      if (child.getName().endsWith("xml")) {
+        readCastVoteRecord(castVoteRecords, child.toPath());
+      }
+    }
+  }
+
+  // parse Cvr xml file into CastVoteRecord objects and add them to the input List<CastVoteRecord>
+  void readCastVoteRecord(List<CastVoteRecord> castVoteRecords, Path path)
+      throws IOException, CvrParseException {
+    try {
+      Logger.log(Level.INFO, "Reading Hart cast vote record file: %s...", path.getFileName());
+
+      XmlMapper xmlMapper = new XmlMapper();
+      xmlMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+      String xmlString = new String(Files.readAllBytes(path));
+      HartCvrXml xmlCvr = xmlMapper.readValue(xmlString, HartCvrXml.class);
+
+      for (Contest contest : xmlCvr.Contests) {
+        // TODO: use contest Id
+        if (!contest.Name.equals(contestConfig.getContestName())) {
+          continue;
+        }
+        if (contest.Options == null) {
+          // undervoted
+          continue;
+        }
+
+        ArrayList<Pair<Integer, String>> rankings = new ArrayList<>();
+        for (Option option : contest.Options) {
+          if (option.WriteInData != null) {
+            Logger.log(Level.WARNING, "");
+          }
+
+          if (!this.contestConfig.getCandidateCodeList().contains(option.Id)) {
+            Logger.log(
+                Level.SEVERE,
+                "Candidate ID: \"%s\" name: \"%s\" from CVR is not in the config file!",
+                option.Id, option.Name);
+            throw new CvrParseException();
+          }
+
+          // Hart RCV election ranks are indicated by a string read left to right:
+          // each digit corresponds to a rank and is set to 1 if that rank was voted.  For example:
+          // 0101 indicates ranks 2 and 4 are voted for this option (overvote)
+          // 00   indicates no rank was voted
+          // 010  indicates rank 2 was voted
+          for (int rank = 1; rank < option.Value.length() + 1; rank++) {
+            String rankValue = option.Value.substring(rank - 1, rank);
+            if (rankValue.equals("1")) {
+              rankings.add(new Pair<>(rank, option.Id));
+            }
+          }
+        }
+
+        CastVoteRecord cvr = new CastVoteRecord(
+            contest.Id,
+            null,
+            xmlCvr.BatchNumber,
+            xmlCvr.CvrGuid,
+            xmlCvr.PrecinctSplit.Name,
+            xmlCvr.PrecinctSplit.Id,
+            null,
+            rankings);
+        castVoteRecords.add(cvr);
+
+        // provide some user feedback on the Cvr count
+        if (castVoteRecords.size() % 50000 == 0) {
+          Logger.log(Level.INFO, "Parsed %d cast vote records.", castVoteRecords.size());
+        }
+      }
+    } catch (Exception e) {
+      Logger.log(Level.SEVERE, "Error parsing cast vote record:\n%s", e.toString());
+      throw e;
+    }
+  }
+
+  static class WriteInData {
```

Comment:
So this and all the classes below are just matching their XML schema so that we can parse it?

Reply:
Yeah - it worked quite well.